src/App.tsx

```diff
@@ -1,223 +1,247 @@
+  const [tempTodoTitle, setTempTodoTitle] = useState('');
+  const [
+    statusToFilter,
+    setStatusToFilter,
+  ] = useState<TodosFilter>(TodosFilter.All);
+
+  const tempTodo = useMemo(() => ({
+    id: 0,
+    userId: 0,
+    title: tempTodoTitle,
+    completed: false,
+  }), [tempTodoTitle]);
+
+  const closeNotification = useCallback(() => setHasError(false), []);
+
+  const filtredTodos = useMemo(() => (
+    todos.filter(({ completed }) => {
+      switch (statusToFilter) {
+        case TodosFilter.Active:
+          return !completed;
+
+        case TodosFilter.Completed:
+          return completed;
+
+        default:
+          return true;
+      }
+    })
+  ), [todos, statusToFilter]);
+
+  const completedTodos = useMemo(() => (
+    todos.filter((todo) => todo.completed)
+  ), [todos]);
+
+  const generateError = useCallback((message: string) => {
+    setErrorMessage(message);
+    setHasError(true);
+  }, []);
+
+  const getTodosFromServer = useCallback(async () => {
+    try {
+      if (user) {
+        const todosFromServer = await getTodos(user.id);
+
+        setTodos(todosFromServer);
+      }
+    } catch (error) {
+      generateError('Unable to show todos');
+    }
+  }, []);
+
+  const addTodoToServer = useCallback(async (todoTitle: string) => {
+    if (user) {
+      try {
+        setIsAdding(true);
+        setTempTodoTitle(todoTitle);
+
+        await addTodo({
+          title: todoTitle,
+          userId: user.id,
+          completed: false,
+        });
+
+        await getTodosFromServer();
+        setIsAdding(false);
+      } catch (error) {
+        generateError('Unable to add todo');
+        setIsAdding(false);
+      }
+    }
+  }, []);
+
+  const removeTodoFromServer = useCallback(async (todoId: number) => {
+    try {
+      setTodoIdsLoading(currIds => [...currIds, todoId]);
+
+      await removeTodo(todoId);
+      await getTodosFromServer();
+
+      setTodoIdsLoading(currIds => (
+        currIds.filter((id) => id !== todoId)
+      ));
+    } catch (error) {
+      generateError(`Unable to remove todo ${todoId}`);
+    }
+  }, []);
+
+  const removeAllCompletedTodos = useCallback(async () => {
+    try {
+      await Promise.all(completedTodos.map(({ id }) => (
+        removeTodoFromServer(id)
+      )));
+    } catch (error) {
+      generateError('Unable to remove all completed todo');
+    }
+  }, [completedTodos]);
+
+  const toggleTodoServerStatus = useCallback(async (
+    todoId: number,
+    status: boolean,
+  ) => {
+    try {
+      setTodoIdsLoading(currIds => [...currIds, todoId]);
+
+      await patchTodo(todoId, { completed: status });
+      await getTodosFromServer();
+
+      setTodoIdsLoading(currIds => (
+        currIds.filter((id) => id !== todoId)
+      ));
+    } catch (error) {
+      generateError(`Unable to change status of todo #${todoId}, todo not exist!`);
+      await getTodosFromServer();
```

Comment:
do we need to load todos from server in case of error?

Reply:
I mean, in this case nothing is changed and there is no need to update current todo list

Reply:
Let's do this only when actions with todos was successful